src/Raven.Server/Documents/Includes/Sharding/IncludeDocumentsOrchestratedSubscriptionCommand.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System.Collections.Generic;
+using System.Threading;
+using Raven.Client.Documents.Session;
+using Raven.Client.Extensions;
+using Raven.Client.Util;
+using Raven.Server.Documents.Queries.Sharding;
+using Raven.Server.Documents.Sharding;
+using Raven.Server.Documents.Sharding.Queries;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Includes.Sharding;
+
+public class IncludeDocumentsOrchestratedSubscriptionCommand : IncludeDocumentsCommand
+{
+    private readonly ClusterOperationContext _clusterContext;
+    private readonly ShardedDatabaseContext _databaseContext;
+    private readonly CancellationToken _token;
+    private List<Document> _result;
+
+    public IncludeDocumentsOrchestratedSubscriptionCommand(ClusterOperationContext clusterContext, ShardedDatabaseContext databaseContext, CancellationToken token)
+    {
+        _clusterContext = clusterContext;
+        _databaseContext = databaseContext;
+        _token = token;
+    }
+
+    internal override bool HasIncludesIds() => _result is { Count: > 0 };
+
+    public override void Fill(List<Document> result, bool includeMissingAsNull)
+    {
+        if (_result == null)
+            return;
+
+
+        foreach (var res in _result)
+        {
+            result.Add(res);
+        }
+
+    }
+
+    public override void Gather(List<BlittableJsonReaderObject> includes)
+    {
+        if (includes == null || includes.Count == 0)
+            return;
+
+        HashSet<string> missingDocumentIncludes = null;
+
+        foreach (var include in includes)
+        {
+            _token.ThrowIfCancellationRequested();
+            using (include)
+            {
+                var prop = new BlittableJsonReaderObject.PropertyDetails();
+                include.GetPropertyByIndex(0, ref prop);
+                if (InMemoryDocumentSessionOperations.TryGetIncludedDocument(prop, out DocumentInfo newDocumentInfo) && newDocumentInfo != null)
+                    continue;
+                if (newDocumentInfo == null)
+                {
+                    // missing include from shard
+                    (missingDocumentIncludes ??= new HashSet<string>()).Add(prop.Name);
+                }
+                else
+                {
+                    PopulateResultWithIncludedDocument(newDocumentInfo, prop);
+                }
+            }
+        }
+
+        if (missingDocumentIncludes == null)
+            return;
+
+        using (_databaseContext.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext ctx))
+        {
+            var queryResult = new ShardedQueryResult();
+            AsyncHelpers.RunSync(() =>
```

Comment:
can we make it `async` all the way?

Reply:
no it will require changing `GetBatch` to be `IAsyncEnumerable` we can open separate issue for that